### PR TITLE
Make format command's filepath optional

### LINKED ISSUE #2008: Add support to read from stdin for the `juvix format` command
At present, the `juvix format` requires a file path or a directory as input for formatting Juvix code. The primary motivation behind this option is to substitute the `dev scope` call in the VSCode extension, enabling us to invoke the `format` command more flexibly and read directly from stdin. Also, notice that this flag `--stdin` is offered as global, so one really expects this works.

### PR DESCRIPTION
**Pull request by vrom911**
# Description

No the filepath in the `juvix forma` command is n=made optional. However, in that case, the `--stdin` command is required.

### Implementation details

~For now, as a quick solution, I have introduce the "fake" path that is used for `fomat` command with stdin option.~
I needed to do a couple of big changes:
 * `format` command FILE is now optional, howvere, I check that in case of `Nothing` `--stdin` option should be present, otherwise it will fail
 * `entryPointModulePaths` is now `[]` instead of `NonEmpty`
 * `ScopeEff` now has `ScopeStdin` constructor as well, which would take the input from stdin instead of having path passed around
 * `RunPipelineNoFileEither` is added to the `App` with the bunch of `*Stdin` functions that doesn't require filepath argument to be passed

Fixes #2008

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works:
  - [x] smoke tests

### CODE REVIEW

File: src/Juvix/Parser/Error.hs [hunk 1]
```diff
@@ -12,13 +13,15 @@ data ParserError
   = ErrMegaparsec MegaparsecError
   | ErrTopModulePath TopModulePathError
   | ErrWrongTopModuleName WrongTopModuleName
+  | ErrStrinOrFile StdinOrFileError
```
Comment (paulcadman) [inline src/Juvix/Parser/Error.hs:16]: typo
```suggestion
  | ErrStdinOrFile StdinOrFileError
```

File: src/Juvix/Parser/Error.hs [hunk 2]
```diff
@@ -12,13 +13,15 @@ data ParserError
...
   deriving stock (Show)
 
 instance ToGenericError ParserError where
   genericError = \case
     ErrMegaparsec e -> genericError e
     ErrTopModulePath e -> genericError e
     ErrWrongTopModuleName e -> genericError e
+    ErrStrinOrFile e -> genericError e
```
Comment (paulcadman) [inline src/Juvix/Parser/Error.hs:24]:
```suggestion
    ErrStdinOrFile e -> genericError e
```

File: src/Juvix/Compiler/Concrete/Translation/FromSource.hs [hunk 3]
```diff
@@ -28,10 +34,22 @@ fromSource ::
...
+  (_resultTable, _resultModules) <- runParserInfoTableBuilder (runReader e getParsedModuleTops)
   let _resultEntry = e
   return ParserResult {..}
   where
+    getParsedModuleTops ::
+      forall r.
+      (Members '[PathResolver, Files, Error ParserError, InfoTableBuilder, NameIdGen] r) =>
+      Sem r (NonEmpty (Module 'Parsed 'ModuleTop))
+    getParsedModuleTops = case (e ^. entryPointStdin, e ^. entryPointModulePaths) of
+      (Nothing, []) -> throw $ ErrStrinOrFile StdinOrFileError
```
Comment (paulcadman) [inline src/Juvix/Compiler/Concrete/Translation/FromSource.hs:46]:
```suggestion
      (Nothing, []) -> throw $ ErrStdinOrFile StdinOrFileError
```

File: app/Commands/Format.hs [hunk 4]
```diff
@@ -17,17 +18,32 @@ data FormatRenderMode
...
-      Right p -> formatProject p
+      Just (Left p) -> format p
+      Just (Right p) -> formatProject p
+      Nothing ->
+        if isStdin
+          then formatStdin
+          else do
+            printFailureExit $
+              Text.unlines
+                [ "juvix format error: either 'JUVIX_FILE_OR_PROJECT' or 'stdin' option should be specified",
```
Comment (paulcadman) [inline app/Commands/Format.hs:43]:
```suggestion
                [ "juvix format error: either 'JUVIX_FILE_OR_PROJECT' or '--stdin' option must be specified",
```

File: app/Commands/Format.hs [hunk 5]
```diff
@@ -17,17 +18,32 @@ data FormatRenderMode
...
+                  "Use --help to learn more usage information."
```
Comment (paulcadman) [inline app/Commands/Format.hs:44]:
```suggestion
                  "Use the --help option to display more usage information."
```

File: app/Commands/Format.hs [hunk 6]
```diff
@@ -17,17 +18,32 @@ data FormatRenderMode
...
+        Just (Right {}) -> TargetDir
+        Nothing -> TargetStdin
   runOutputSem (renderFormattedOutput target opts) $ runScopeFileApp $ do
     res <- case f of
-      Left p -> format p
-      Right p -> formatProject p
+      Just (Left p) -> format p
+      Just (Right p) -> formatProject p
+      Nothing ->
+        if isStdin
```
Comment (paulcadman) [inline app/Commands/Format.hs:38]: Style: We prefer to use multiway-if rather than `if ... then ... else`.

i.e change to

```
if 
   | isStdin -> ...
   | otherwise -> ...
```

Comment (vrom911) [reply app/Commands/Format.hs]: Good to know! Changed to that 🙂 
Do we have a style guide on all these things I can check?

Comment (paulcadman) [reply app/Commands/Format.hs]: Unfortunately we do not have anything written down. I think we should, as long as we make it clear that we always expect exceptions tot he guidelines. Let's discuss with the whole team.

Comment (vrom911) [reply app/Commands/Format.hs]: Sounds great!

**Review by paulcadman — APPROVED** [review]
Nicely done 🌟

**Review by janmasrovira — COMMENTED** [review]
I have some suggestions

File: src/Juvix/Formatter.hs [hunk 7]
```diff
@@ -98,6 +93,38 @@ formatProject p = do
...
+
+formatResultFromContents ::
+  forall r.
+  Members '[Output FormattedFileInfo] r =>
+  Text ->
+  NonEmpty AnsiText ->
+  Path Abs File ->
+  Sem r FormatResult
+formatResultFromContents originalContents formattedContents filepath
+  | originalContents /= (ansiPlainText formattedContents) = do
```
[the remark this answers is not in the document]
Comment (vrom911) [reply src/Juvix/Formatter.hs:116]: Yess!
Do we run `hlint` somewhere?
Maybe it would make sense to add ti to `pre-commit`?

File: src/Juvix/Prelude/Lens.hs [hunk 8]
```diff
@@ -6,6 +6,19 @@ import Juvix.Prelude.Base
 _head1 :: Lens' (NonEmpty a) a
 _head1 = singular each
 
+-- | points to the first element of a the list.
```
[the remark this answers is not in the document]
Comment (vrom911) [reply src/Juvix/Prelude/Lens.hs:9]: True, I can use `_head` lens, but for implementin the lens like:
```haskell
mainModulePath :: Lens' EntryPoint (Maybe (Path Abs File))
mainModulePath = entryPointModulePaths . _headMaybe
```

will still need to write it manually.
Would you suggest still to use `^?` and `_head` everywhere else?

File: app/App.hs [hunk 9]
```diff
@@ -6,7 +6,9 @@ import GlobalOptions
 import Juvix.Compiler.Concrete.Translation.FromParsed.Analysis.PathResolver
 import Juvix.Compiler.Pipeline
 import Juvix.Data.Error qualified as Error
-import Juvix.Prelude.Pretty hiding (Doc)
+import Juvix.Prelude.Pretty hiding
+  ( Doc,
+  )
```
Comment (jonaprieto) [inline app/App.hs:11]: revert this change. less diff as much as we can.

File: app/Commands/Dev/Geb/Repl.hs [hunk 10]
```diff
@@ -2,7 +2,9 @@
 
 module Commands.Dev.Geb.Repl where
 
-import Commands.Base hiding (command)
+import Commands.Base hiding
+  ( command,
+  )
```
Comment (jonaprieto) [inline app/Commands/Dev/Geb/Repl.hs:7]: Idem.

File: app/App.hs (hunk 9 above)
Comment (vrom911) [reply app/App.hs:11]: I am pretty sure this is ormolu, but let me double check

Comment (vrom911) [reply app/App.hs:11]: Yes, this is the result of the `make ormolu` command for me

File: app/GlobalOptions.hs [hunk 11]
```diff
@@ -150,6 +150,9 @@ entryPointFromGlobalOptionsPre roots premainFile opts = do
   mainFile <- prepathToAbsFile (roots ^. rootsInvokeDir) premainFile
   entryPointFromGlobalOptions roots mainFile opts
 
+entryPointFromGlobalOptionsNoFilePre :: Roots -> GlobalOptions -> IO EntryPoint
+entryPointFromGlobalOptionsNoFilePre roots opts = entryPointFromGlobalOptionsNoFile roots opts
```
Comment (jonaprieto) [inline app/GlobalOptions.hs:154]: Do you really need to define this?

File: app/Commands/Dev/Geb/Repl.hs [hunk 12]
```diff
@@ -60,17 +62,21 @@ loadEntryPoint ep = do
...
-  let evalRes =
-        Geb.runEval $
-          Geb.RunEvalArgs
-            { _runEvalArgsContent = content,
-              _runEvalArgsInputFile = epPath,
-              _runEvalArgsEvaluatorOptions = Geb.defaultEvaluatorOptions
-            }
-  printEvalResult evalRes
+  let epPath :: Maybe (Path Abs File) = ep ^. entryPointModulePaths . _headMaybe
+  case epPath of
```
[the remark this answers is not in the document]
Comment (vrom911) [reply app/Commands/Dev/Geb/Repl.hs]: Done 👍🏼

File: src/Juvix/Prelude/Lens.hs (hunk 8 above)
[the remark this answers is not in the document]
Comment (janmasrovira) [reply src/Juvix/Prelude/Lens.hs:9]: That's true, I think we should have a `Traversal'` instead. 
```
mainModulePath :: Traversal' EntryPoint (Path Abs File)
mainModulePath = entryPointModulePaths . _head
```

Then in the call sites we will need to do `entry ^? mainModulePath`

Ideally we would use a `Prism` but microlens doesn't have them

File: src/Juvix/Compiler/Pipeline.hs [hunk 13]
```diff
@@ -192,6 +192,8 @@ upToEval =
 -- Internal workflows
 --------------------------------------------------------------------------------
 
+instance Member Parser.PragmasStash '[]
```
[the remark this answers is not in the document]
Comment (vrom911) [reply src/Juvix/Compiler/Pipeline.hs]: Thanks! This makes sense 👍🏼

File: app/App.hs (hunk 9 above)
Comment (jonaprieto) [reply app/App.hs:11]: Which version of ormolu are you running? The CI runs "ormolu 0.5.2.0 HEAD 810eb7839cbf4e358821447e8f4603955999e62d using ghc-lib-parser 9.4.2.20220822"

Comment (vrom911) [reply app/App.hs:11]: I'm using `0.5.2.0` too:
Screenshot 2023-04-27 at 12 44 19

**Review by jonaprieto — APPROVED** [review]

**Comment by jonaprieto** [comment]
Good job
